lib/Dialect/Arc/Transforms/SinkInputs.cpp

```diff
@@ -0,0 +1,134 @@
+//===- SinkInputs.cpp------ -----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-sink-inputs"
+
+using namespace circt;
+using namespace arc;
+using namespace hw;
+
+namespace {
+struct SinkInputsPass : public SinkInputsBase<SinkInputsPass> {
+  void runOnOperation() override;
+  void runOnModule();
+};
+} // namespace
+
+void SinkInputsPass::runOnOperation() {
+  // For each arc, this map tracks which arguments are constant.
+  DenseMap<StringRef, SmallVector<ConstantOp>> arcConstArgs;
+  // All stateOps that use a particular arc.
+  DenseMap<StringRef, SmallVector<StateOp>> arcUses;
+
+  auto module = getOperation();
+  for (auto moduleOp :
+       llvm::make_early_inc_range(module.getOps<HWModuleOp>())) {
+    for (auto op : moduleOp.getBody().getOps<StateOp>()) {
+      // Find all stateOps that use constant operands.
+      auto ops = op.getInputs();
+      SmallVector<ConstantOp> constArgs(ops.size());
+      for (unsigned i = 0; i < ops.size(); i++) {
+        auto arg = ops[i];
+        if (auto c = dyn_cast_or_null<ConstantOp>(arg.getDefiningOp()))
+          constArgs[i] = c;
+      }
+      auto arc = op.getArc();
+      arcUses[arc].push_back(op);
+      if (arcConstArgs.count(arc) == 0) {
+        arcConstArgs[arc] = constArgs; // make an entry for this arc
+      } else {
+        SmallVector<ConstantOp> merged(ops.size());
+        auto args = arcConstArgs[arc];
+        for (unsigned i = 0; i < args.size(); i++) {
+          if (args[i] == constArgs[i])
+            merged[i] = constArgs[i];
+          else
+            merged[i] = nullptr;
+        }
+        arcConstArgs[arc] = merged;
+      }
+    }
+  }
+
+  // Now we go through all the defines and move the constant ops into the
+  // bodies and rewrite the function types.
+  for (auto defOp : llvm::make_early_inc_range(module.getOps<DefineOp>())) {
+    if (arcConstArgs.count(defOp.getName()) >= 1) {
+      OpBuilder builder = OpBuilder::atBlockBegin(&defOp.getBodyBlock());
+      auto args = arcConstArgs[defOp.getName()];
+      DenseSet<unsigned> deletedArgs;
```

Comment:
Can't we just index into the `toDelete` bitvector instead of maintaining this as well?